vaadin-radio-button-flow-parent/vaadin-radio-button-flow/src/test/java/com/vaadin/flow/component/radiobutton/RadioButtonGroupTest.java

```diff
@@ -417,4 +424,63 @@ public void addNullOption_setValue() {
         group.setValue(null);
         Assert.assertEquals(group.getValue(), null);
     }
+
+    @Test
+    public void dataProviderListeners_radioButtonGroupAttached_oldDataProviderListenerRemoved() {
+        AtomicInteger listenersCount = new AtomicInteger(0);
+        AtomicBoolean oldListenerRemoved = new AtomicBoolean();
+
+        // given
+        RadioButtonGroup<String> rbg = new RadioButtonGroup<>();
+        ListDataProvider<String> dataProvider = new ListDataProvider<String>(
+                Collections.emptyList()) {
+            @Override
+            public Registration addDataProviderListener(
+                    DataProviderListener<String> listener) {
+                listenersCount.incrementAndGet();
+                Registration registration = super.addDataProviderListener(
+                        listener);
+                // the first listener is added by RBG in 'setDataProvider'
+                // and it is the one we want to be removed.
+                return listenersCount.get() == 1
+                        ? Registration.combine(registration,
+                                () -> oldListenerRemoved.set(true))
+                        : registration;
+            }
+        };
+
+        // when
+        rbg.setDataProvider(dataProvider);
+
+        // then
+        Assert.assertEquals(
+                "Expected exactly one data provider listener added by RBG's "
+                        + "'setDataProvider' method",
+                1, listenersCount.get());
+        Assert.assertFalse(
+                "Expected data provider listener added by RBG not to be "
+                        + "removed before RBG being attached",
+                oldListenerRemoved.get());
+
+        // when
+        DataCommunicatorTest.MockUI mockUI = new DataCommunicatorTest.MockUI();
+        mockUI.add(rbg);
+        fakeClientCommunication(mockUI);
+
+        // then
+        Assert.assertEquals(
+                "Expected exactly one data provider listener added by "
+                        + "RBG's attach handle method",
+                2, listenersCount.get());
```

Comment:
Again, it verifies for against two but there should be only one ?

Reply:
`listenersCount` will be re-set so as to validate agains expected '1' value, not '2'.